docs/design/proposals/creds-from-kms.md

```diff
@@ -0,0 +1,131 @@
+# Fetching Ceph user credentials from KMS
+
+## Proposal
+
+Ceph-CSI supports several KMS implementations for storing and
+fetching DEKs (Data Encryption Keys) used for RBD volume encryption.
+The focus of this proposal is to leverage existing KMS implementation
+to support obtaining Ceph user keyring used while
+creating, deleting, mapping and resizing volumes.
+
+### Benefits
+
+- For scenarios where Ceph is deployed externally, either as
+  a standalone or through Rook, one needs to obtain the Ceph keyring(s)
+  and manually create the K8s secrets needed by Ceph-CSI. With this feature,
+  sensitive keyring(s) can be stored securely in an
+  external secret management system such as Hashicorp Vault,
+  AWS Secret Manager, etc., and have Ceph-CSI pull these
+  keys directly from them.
+- More secure compared to K8s secrets,
+  which are relatively easy to access and
+  store sensitive secrets as unencrypted based64 encoded text.
+
+### Drawbacks
+
+- Adds additional overhead as on each CSI RPC call
+  a request is made to KMS for fetching credentials.
+- Possible risk of hitting the KMS rate limit.
+
+## Extending existing implementation
+
+- KMS ID is provided as part of the provisioner,
+  node-stage, and controller-expand secrets, along with Ceph `userID`, as shown below.
+  Having this information as part of *StorageClass* parameters,
+  like using the `encryptionKMSID` key for volume encryption,
+  wouldn't work as these parameters are not passed
+  to all CSI RPCs, e.g. `DeleteVolume`, `ControllerExpand` etc.
```

Comment:
The parameters in the StorageClass are copied into the PersistentVolume.CSIDriver parameters. I think it would be cleaner to not use the common secrets in the StorageClass for this, but directly refer to a `kmsID` with potentially additional parameters like `userID`.

Reply:
It is not very clear to me on how we can obtain the `PersistentVolume.CSIDriver`  parameters from calls like `DeleteVolume`. 

Looking at the DeleteVolume method and DeleteVolumeRequest struct:

```golang
func (cs *ControllerServer) DeleteVolume(
	ctx context.Context,
	req *csi.DeleteVolumeRequest,
) (*csi.DeleteVolumeResponse, error)

// ....

type DeleteVolumeRequest struct {
	// The ID of the volume to be deprovisioned.
	// This field is REQUIRED.
	VolumeId string `protobuf:"bytes,1,opt,name=volume_id,json=volumeId,proto3" json:"volume_id,omitempty"`
	// Secrets required by plugin to complete volume deletion request.
	// This field is OPTIONAL. Refer to the `Secrets Requirements`
	// section on how to use this field.
	Secrets              map[string]string `protobuf:"bytes,2,rep,name=secrets,proto3" json:"secrets,omitempty" protobuf_key:"bytes,1,opt,name=key,proto3" protobuf_val:"bytes,2,opt,name=value,proto3"`
	XXX_NoUnkeyedLiteral struct{}          `json:"-"`
	XXX_unrecognized     []byte            `json:"-"`
	XXX_sizecache        int32             `json:"-"`
}
```

I only see that we can get VolumeId as part of this request. Are you suggesting we get the `PersistentVolume` object here obtain these parameters? If yes, wouldn't that require an additional API call?

Do you mind helping me understand this better?